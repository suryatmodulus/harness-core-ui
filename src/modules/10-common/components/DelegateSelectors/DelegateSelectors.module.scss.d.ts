/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly wrapper: string
  readonly loader: string
  readonly delegatePopover: string
  readonly delegateInput: string
}
export default styles
