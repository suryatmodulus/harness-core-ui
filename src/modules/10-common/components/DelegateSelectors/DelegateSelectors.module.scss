.wrapper {
  position: relative;
  flex-shrink: 1 !important;

  .loader {
    display: flex;
    height: 32px;
    background: var(--white);
    align-items: center;
    border: 1px solid var(--form-field-border);
    border-radius: 4px;
    padding: 0 var(--spacing-small);
    column-gap: var(--spacing-small);

    & :global(.bp3-icon) {
      margin: 0 !important;
    }
  }
}

.delegatePopover {
  position: relative;

  :global(.bp3-input) {
    display: block;
    border: 1px solid var(--grey-400);
    border-radius: 4px;
    font-size: var(--font-size-normal);
    padding: 0 var(--spacing-small);
    min-height: 32px;
    color: var(--grey-800);
    box-shadow: none;
  }

  & :global(.bp3-transition-container),
  & :global(.bp3-popover) {
    width: 100%;
    box-shadow: none;
    border-radius: 0 0 4px 4px;
  }

  & :global(.bp3-menu) {
    max-height: 350px;
    width: 100%;
    overflow: auto;
    box-shadow: none;
    border: 1px solid var(--grey-400);
    border-radius: 0 0 4px 4px;
    padding: 0;
  }

  & :global(.bp3-input.bp3-tag-input) {
    max-height: fit-content !important;
  }
}

.delegateInput {
  input,
  :global(.bp3-tag) {
    margin-bottom: 0 !important;
  }
}
