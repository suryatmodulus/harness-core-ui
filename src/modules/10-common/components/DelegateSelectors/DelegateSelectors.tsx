import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import cx from 'classnames'
import { SimpleTagInput, Text, Icon, Color } from '@wings-software/uicore'
import { useToaster } from '@common/exports'
import { useStrings } from 'framework/strings'
import { useGetDelegateSelectorsUpTheHierarchy } from 'services/portal'

import type { AccountPathProps, ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import css from './DelegateSelectors.module.scss'

const isValidExpression = (
  tag: string,
  showError: (message: React.ReactNode, timeout?: number, key?: string) => void,
  errorMsg: string
): boolean => {
  let validExpression = true
  if (tag.includes('${')) {
    validExpression = tag.includes('${') && tag.includes('}')
    if (!validExpression) {
      showError(errorMsg, 5000)
    }
  }
  return validExpression
}

export interface DelegateSelectorsProps
  extends Partial<React.ComponentProps<typeof SimpleTagInput>>,
    Partial<ProjectPathProps> {
  placeholder?: string
  pollingInterval?: number
  wrapperClassName?: string
}

export const DelegateSelectors = (props: DelegateSelectorsProps): React.ReactElement | null => {
  const { accountId } = useParams<AccountPathProps>()
  const { orgIdentifier, projectIdentifier, pollingInterval = null, wrapperClassName, placeholder, ...rest } = props

  const { getString } = useStrings()
  const { showError } = useToaster()

  const queryParams = { accountId, orgId: orgIdentifier, projectId: projectIdentifier }
  const {
    data: apiData,
    loading,
    refetch
  } = useGetDelegateSelectorsUpTheHierarchy({
    queryParams
  })

  const [data, setData] = useState(apiData)

  useEffect(() => {
    if (apiData) {
      setData(apiData)
    }
  }, [apiData])

  // polling logic
  useEffect(() => {
    if (pollingInterval === null) {
      return
    }
    let id: number | null
    if (!loading) {
      id = window.setTimeout(() => refetch(), pollingInterval)
    }
    return () => {
      if (id) {
        window.clearTimeout(id)
      }
    }
  }, [data, loading, refetch, pollingInterval])

  return (
    <div className={cx(css.wrapper, wrapperClassName)} data-name="DelegateSelectors">
      {loading && !data ? (
        <div className={css.loader}>
          <Icon margin="medium" name="spinner" size={15} color={Color.PRIMARY_8} />
          <span>{getString('loading')}</span>
        </div>
      ) : (
        <SimpleTagInput
          fill
          popoverProps={{
            usePortal: false,
            minimal: true,
            position: 'bottom-left',
            className: css.delegatePopover
          }}
          items={data?.resource || []}
          {...rest}
          allowNewTag
          getTagProps={(value, _index, _selectedItems, createdItems, items) => {
            const _value = value as string
            const isItemNewlyCreated = createdItems.includes(_value) || !items.includes(_value)
            const isExpression = isItemNewlyCreated && _value.startsWith('${') && _value.endsWith('}')
            return isExpression
              ? { intent: 'none', minimal: true }
              : isItemNewlyCreated
              ? { intent: 'danger', minimal: true }
              : { intent: 'primary', minimal: true }
          }}
          validateNewTag={(tag: string) => {
            const pattern = new RegExp('^[a-z0-9-${}]+$', 'i')
            const validTag = new RegExp('^[a-z0-9-${}._<>+]+$', 'i').test(tag)
            const tagChars = tag.split('')
            const validExpression = isValidExpression(
              tag,
              showError,
              getString('delegate.DelegateSelectorErrorMessage')
            )
            const invalidChars = new Set()

            if (!validTag) {
              const errorMsg = (
                <Text>
                  {getString('delegate.DelegateSelector')}
                  <>
                    {tagChars.map((item: string) => {
                      if (!pattern.test(item)) {
                        invalidChars.add(item)
                        return <strong style={{ color: 'red' }}>{item}</strong>
                      } else {
                        return item
                      }
                    })}
                  </>
                  {getString('delegate.DelegateSelectorErrMsgSplChars')}: {Array.from(invalidChars).join(',')}
                </Text>
              )
              showError(errorMsg, 5000)
            }
            return validTag && validExpression
          }}
          placeholder={placeholder || getString('delegate.Delegate_Selector_placeholder')}
          className={css.delegateInput}
        />
      )}
    </div>
  )
}

export default DelegateSelectors
