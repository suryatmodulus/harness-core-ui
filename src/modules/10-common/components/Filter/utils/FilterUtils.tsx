/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { isEmpty, isUndefined, omitBy } from 'lodash-es'
import React from 'react'
import type { FilterDTO } from 'services/cd-ng'

import css from './FilterUtils.module.scss'

type supportedTypes = string | number | boolean | Record<string, any>

const tagSeparator = ':'

interface NGTag {
  key: string
  value: string
}

export const renderItemByType = (data: supportedTypes | Array<supportedTypes> | Record<string, any>): string => {
  if (Array.isArray(data)) {
    return data?.map(item => renderItemByType(item)).join(', ')
  } else if (typeof data === 'object') {
    if (Object.prototype.hasOwnProperty.call(data, 'key') && Object.prototype.hasOwnProperty.call(data, 'value')) {
      const { key, value } = data as NGTag
      return key.toString().concat(value ? tagSeparator.concat(value.toString()) : '')
    }
    return Object.entries(data)
      .map(([key, value]) => {
        return key.toString().concat(value ? tagSeparator.concat(value.toString()) : '')
      })
      .join(', ')
  } else if (typeof data === 'number') {
    return data.toString()
  } else if (typeof data === 'boolean') {
    return data ? 'true' : 'false'
  }
  return typeof data === 'string' ? data : ''
}

export const getFilterSummary = (
  fieldToLabelMapping: Map<string, string>,
  fields: Record<string, any>
): JSX.Element => {
  return (
    <ol className={css.noStyleUl}>
      {Object.entries(fields as Record<string, any>).map(([key, value]) => {
        if (fieldToLabelMapping.has(key)) {
          return (
            <li key={key} className={css.summaryItem}>
              <span>{fieldToLabelMapping.get(key)} : </span>
              <span style={{ fontWeight: 'bold' }}>{renderItemByType(value)}</span>
            </li>
          )
        }
      })}
    </ol>
  )
}

export const removeNullAndEmpty = (object: Record<string, any>) => omitBy(omitBy(object, isUndefined), isEmpty)
export const isObjectEmpty = (arg: Record<string, any>): boolean => isEmpty(omitBy(arg, isEmpty))
export const UNSAVED_FILTER = 'Unsaved Filter'
export const MAX_FILTER_NAME_LENGTH = 25

export const flattenObject = (obj: Record<string, any>) => {
  const flattened: {
    [key: string]: string
  } = {}
  Object.keys(obj).forEach((key: string) => {
    if (Object.prototype.toString.call(obj[key]) === '[object Object]' && obj[key] !== null) {
      Object.assign(flattened, flattenObject(obj[key]))
    } else {
      flattened[key as string] = obj[key]
    }
  })
  return flattened
}

export const getFilterSize = (obj: Record<string, any>): number => {
  return Object.keys(flattenObject(obj)).length
}

export const getFilterByIdentifier = (filters: FilterDTO[], identifier: string): FilterDTO | undefined =>
  filters?.find((filter: FilterDTO) => filter.identifier?.toLowerCase() === identifier.toLowerCase())
