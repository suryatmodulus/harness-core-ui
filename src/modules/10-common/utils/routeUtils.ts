/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type {
  AccountPathProps,
  OrgPathProps,
  ProjectPathProps,
  PipelinePathProps,
  TriggerPathProps,
  ExecutionPathProps,
  ConnectorPathProps,
  SecretsPathProps,
  PipelineType,
  FeatureFlagPathProps,
  SegmentPathProps,
  CVDataSourceTypePathProps,
  BuildPathProps,
  EnvironmentPathProps,
  DelegatePathProps,
  DelegateConfigProps,
  InputSetPathProps,
  VerificationPathProps,
  TargetPathProps,
  ModulePathParams,
  RolePathProps,
  ResourceGroupPathProps,
  UserGroupPathProps,
  UserPathProps,
  ServiceAccountPathProps,
  ServicePathProps,
  TemplateStudioPathProps
} from '@common/interfaces/RouteInterfaces'

export const accountPathProps: AccountPathProps = {
  accountId: ':accountId'
}

export const orgPathProps: OrgPathProps = {
  ...accountPathProps,
  orgIdentifier: ':orgIdentifier'
}

export const projectPathProps: ProjectPathProps = {
  ...orgPathProps,
  projectIdentifier: ':projectIdentifier'
}

export const pipelinePathProps: PipelinePathProps = {
  ...projectPathProps,
  pipelineIdentifier: ':pipelineIdentifier'
}

export const templatePathProps: TemplateStudioPathProps = {
  ...projectPathProps,
  templateIdentifier: ':templateIdentifier',
  templateType: ':templateType'
}

export const inputSetFormPathProps: InputSetPathProps = {
  ...pipelinePathProps,
  inputSetIdentifier: ':inputSetIdentifier'
}

export const triggerPathProps: TriggerPathProps = {
  ...pipelinePathProps,
  triggerIdentifier: ':triggerIdentifier'
}

export const executionPathProps: ExecutionPathProps = {
  ...pipelinePathProps,
  executionIdentifier: ':executionIdentifier'
}

export const connectorPathProps: ConnectorPathProps = {
  connectorId: ':connectorId'
}
export const verificationPathProps: VerificationPathProps = {
  verificationId: ':verificationId'
}

export const secretPathProps: SecretsPathProps = {
  secretId: ':secretId'
}

export const rolePathProps: RolePathProps = {
  roleIdentifier: ':roleIdentifier'
}

export const userGroupPathProps: UserGroupPathProps = {
  userGroupIdentifier: ':userGroupIdentifier'
}

export const userPathProps: UserPathProps = {
  userIdentifier: ':userIdentifier'
}

export const serviceAccountProps: ServiceAccountPathProps = {
  serviceAccountIdentifier: ':serviceAccountIdentifier'
}

export const resourceGroupPathProps: ResourceGroupPathProps = {
  resourceGroupIdentifier: ':resourceGroupIdentifier'
}

export const delegatePathProps: DelegatePathProps = {
  delegateIdentifier: ':delegateIdentifier'
}

export const delegateConfigProps: DelegateConfigProps = {
  delegateConfigIdentifier: ':delegateConfigIdentifier'
}

export const modulePathProps: ModulePathParams = {
  module: ':module'
}

export const pipelineModuleParams: Record<keyof PipelineType<unknown>, 'ci' | 'cd' | 'cf' | ':module'> = {
  module: ':module'
}

export const featureFlagPathProps: FeatureFlagPathProps = {
  featureFlagIdentifier: ':featureFlagIdentifier'
}

export const cvDataSourceTypePathProps: CVDataSourceTypePathProps = {
  dataSourceType: ':dataSourceType'
}

export const buildPathProps: BuildPathProps = {
  ...projectPathProps,
  buildIdentifier: ':buildIdentifier'
}

export const environmentPathProps: EnvironmentPathProps = {
  environmentIdentifier: ':environmentIdentifier'
}

export const segmentPathProps: SegmentPathProps = {
  segmentIdentifier: ':segmentIdentifier'
}

export const targetPathProps: TargetPathProps = {
  targetIdentifier: ':targetIdentifier'
}

export const servicePathProps: ServicePathProps = {
  serviceId: ':serviceId'
}

export function withAccountId<T>(fn: (args: T) => string) {
  return (params: T & { accountId: string }): string => {
    const path = fn(params)

    return `/account/${params.accountId}/${path.replace(/^\//, '')}`
  }
}

export function withOrgIdentifier<T>(fn: (args: T) => string) {
  return (params: T & { orgIdentifier: string }): string => {
    const path = fn(params)

    return `/orgs/${params.orgIdentifier}/${path.replace(/^\//, '')}`
  }
}

export function withProjectIdentifier<T>(fn: (args: T) => string) {
  return (params: T & { projectIdentifier: string }): string => {
    const path = fn(params)

    return `/projects/${params.projectIdentifier}/${path.replace(/^\//, '')}`
  }
}

export const getScopeBasedRoute = ({
  scope: { orgIdentifier, projectIdentifier, module },
  path
}: {
  scope: Partial<ProjectPathProps & ModulePathParams>
  path: string
}): string => {
  if (module && orgIdentifier && projectIdentifier) {
    return `/${module}/orgs/${orgIdentifier}/projects/${projectIdentifier}/setup/${path}`
  } else if (orgIdentifier && projectIdentifier) {
    return `/home/orgs/${orgIdentifier}/projects/${projectIdentifier}/setup/${path}`
  } else if (orgIdentifier) {
    return `/settings/organizations/${orgIdentifier}/setup/${path}`
  }
  return `/settings/${path}`
}

export const returnUrlParams = (url: string): string => `?returnUrl=${url}`

export const validateReturnUrl = (url: string): boolean => {
  const decodedUrl = decodeURIComponent(url)

  if (decodedUrl.startsWith('/')) {
    return true
  }

  try {
    const validUrl = new URL(decodedUrl)
    return window.location.hostname === validUrl.hostname
  } catch (_e) {
    return false
  }
}

export const returnLaunchUrl = (url: string): string => `${window.location.pathname.replace(/\/ng\//, '/')}${url}`
