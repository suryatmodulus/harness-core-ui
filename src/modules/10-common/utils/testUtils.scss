/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.browser {
  border: 1px solid #cccccc;
  border-radius: 8px;
  min-height: 300px;
  overflow: auto;

  .browser-header {
    background-color: #cccccc;
    padding: 8px 16px;
    display: flex;
  }

  .browser-path {
    display: block;
    border: none;
    line-height: 14px;
    font-size: 12px;
    color: #444444;
    border-radius: 9px;
    min-width: 100px;
    padding: 2px 8px;
    flex: 1 1 auto;
    border: 1px solid transparent;

    &:focus {
      border-color: rgb(30, 167, 253);
      outline: none;
    }
  }

  .browser-content {
    padding: 8px;
  }
}
