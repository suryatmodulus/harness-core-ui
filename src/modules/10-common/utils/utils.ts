/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}
