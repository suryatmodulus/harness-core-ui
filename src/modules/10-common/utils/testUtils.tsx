/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { queryByAttribute } from '@testing-library/react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation, useHistory } from 'react-router-dom'
import { ModalProvider } from '@wings-software/uicore'
import qs from 'qs'

import { enableMapSet } from 'immer'
import { AppStoreContext, AppStoreContextProps } from 'framework/AppStore/AppStoreContext'
import { LicenseStoreContext, LicenseStoreContextProps } from 'framework/LicenseStore/LicenseStoreContext'
import { LICENSE_STATE_VALUES } from 'framework/LicenseStore/licenseStoreUtil'
import { withAccountId, accountPathProps } from '@common/utils/routeUtils'
import type { Project } from 'services/cd-ng'
import { StringsContext } from 'framework/strings'

import { FeaturesContext, FeaturesContextProps } from 'framework/featureStore/FeaturesContext'
import type { FeatureDetail, FeatureMetaData } from 'framework/featureStore/featureStoreUtil'
import type { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import './testUtils.scss'
import { PermissionsContext, PermissionsContextProps } from 'framework/rbac/PermissionsContext'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseGetMockDataWithMutateAndRefetch<T> extends UseGetMockData<T> {
  mutate: () => Record<string, unknown>
  refetch: () => Record<string, unknown>
  cancel?: () => Record<string, unknown>
}

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): HTMLElement | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): HTMLElement | null => document.querySelector('.bp3-popover-content')

export interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, unknown>
  defaultAppStoreValues?: Partial<AppStoreContextProps>
  defaultLicenseStoreValues?: Partial<LicenseStoreContextProps>
  defaultPermissionValues?: Partial<PermissionsContextProps>
  defaultFeaturesValues?: Partial<FeaturesContextProps>
  projects?: Project[]
  enableBrowserView?: boolean
  stringsData?: Record<string, string>
  getString?(key: string): string
}

export const prependAccountPath = (path: string): string => withAccountId(() => path)(accountPathProps)

export const CurrentLocation = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export interface BrowserViewProps {
  enable?: boolean
  children: React.ReactNode
}

export function BrowserView(props: BrowserViewProps): React.ReactElement {
  const { enable, children } = props
  const location = useLocation()
  const history = useHistory()

  if (!enable) {
    return <>{children}</>
  }

  function handlePathChange(e: React.ChangeEvent<HTMLInputElement>): void {
    history.replace(e.currentTarget.value)
  }

  const search = location.search ? `?${location.search.replace(/^\?/, '')}` : ''

  return (
    <div className="browser">
      <div className="browser-header">
        <input className="browser-path" value={location.pathname + search} onChange={handlePathChange} />
      </div>
      <div className="browser-content">{children}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  enableMapSet()
  const {
    path = '/',
    pathParams = {},
    defaultAppStoreValues,
    queryParams = {},
    defaultLicenseStoreValues,
    defaultPermissionValues,
    defaultFeaturesValues,
    stringsData = {},
    getString = (key: string) => key
  } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const history = React.useMemo(() => createMemoryHistory({ initialEntries: [routePath] }), [])

  const defaultReturn = {
    enabled: true
  }

  /** TODO: Try fixing this later. This is causing some tests to fail */
  // React.useEffect(() => {
  //   history.replace(compile(path)(pathParams) + qs.stringify(queryParams, { addQueryPrefix: true }))
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, [path, pathParams, queryParams])

  return (
    <StringsContext.Provider value={{ data: stringsData as any, getString }}>
      <AppStoreContext.Provider
        value={{
          featureFlags: {
            FEATURE_ENFORCEMENT_ENABLED: true
          },
          updateAppStore: () => void 0,
          currentUserInfo: {},
          ...defaultAppStoreValues
        }}
      >
        <LicenseStoreContext.Provider
          value={{
            versionMap: {},
            licenseInformation: {},
            CI_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            FF_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            CCM_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            CD_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
            updateLicenseStore: () => void 0,
            ...defaultLicenseStoreValues
          }}
        >
          <PermissionsContext.Provider
            value={{
              permissions: new Map<string, boolean>(),
              requestPermission: () => void 0,
              checkPermission: () => true,
              cancelRequest: () => void 0,
              ...defaultPermissionValues
            }}
          >
            <FeaturesContext.Provider
              value={{
                features: new Map<FeatureIdentifier, FeatureDetail>(),
                featureMap: new Map<FeatureIdentifier, FeatureMetaData>(),
                getEdition: () => {
                  return undefined
                },
                requestFeatures: () => void 0,
                requestLimitFeature: () => void 0,
                checkFeature: () => {
                  return defaultReturn
                },
                checkLimitFeature: () => {
                  return defaultReturn
                },
                getRestrictionType: () => {
                  return undefined
                },
                ...defaultFeaturesValues
              }}
            >
              <Router history={history}>
                <ModalProvider>
                  <RestfulProvider base="/">
                    <BrowserView enable={props.enableBrowserView}>
                      <Switch>
                        <Route exact path={path}>
                          {props.children}
                        </Route>
                        <Route>
                          <CurrentLocation />
                        </Route>
                      </Switch>
                    </BrowserView>
                  </RestfulProvider>
                </ModalProvider>
              </Router>
            </FeaturesContext.Provider>
          </PermissionsContext.Provider>
        </LicenseStoreContext.Provider>
      </AppStoreContext.Provider>
    </StringsContext.Provider>
  )
}

export const queryByNameAttribute = (name: string, container: HTMLElement): HTMLElement | null =>
  queryByAttribute('name', container, name)
