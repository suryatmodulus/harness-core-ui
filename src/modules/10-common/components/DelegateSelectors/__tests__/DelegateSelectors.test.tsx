/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import {
  render
  // fireEvent
} from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { accountPathProps } from '@common/utils/routeUtils'
import routes from '@common/RouteDefinitions'
import { DelegateSelectors } from '../DelegateSelectors'

const mockGetCallFunction = jest.fn()
jest.mock('services/portal', () => ({
  useGetDelegateSelectorsUpTheHierarchy: jest.fn().mockImplementation(args => {
    mockGetCallFunction(args)
    return []
  })
}))
const props = {
  urlParams: {
    accountId: 'accountId'
  }
}

const params = {
  accountId: 'testAcc'
}
const TEST_PATH = routes.toDelegates({
  ...accountPathProps
})
describe('Delegate Selectors test', () => {
  test('snapshot testing', () => {
    const { container } = render(
      <TestWrapper path={TEST_PATH} pathParams={params}>
        <DelegateSelectors
          {...props}
          fill
          items={[]}
          {...props}
          allowNewTag
          getTagProps={(value, _index, _selectedItems, createdItems) => {
            return createdItems.includes(value as string) ? { intent: 'danger', minimal: true } : { minimal: true }
          }}
        />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
