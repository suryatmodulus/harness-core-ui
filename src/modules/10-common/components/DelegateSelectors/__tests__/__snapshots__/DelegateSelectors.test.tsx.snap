// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delegate Selectors test snapshot testing 1`] = `
<div>
  <div
    class="wrapper"
    data-name="DelegateSelectors"
  >
    <div
      class="bp3-popover-wrapper delegateInput delegatePopover bp3-fill"
    >
      <div
        class="bp3-popover-target"
      >
        <div
          class=""
        >
          <div
            class="bp3-input bp3-tag-input bp3-fill bp3-multi-select"
          >
            <div
              class="bp3-tag-input-values"
            >
              <input
                class="bp3-input-ghost bp3-multi-select-tag-input-input"
                placeholder="delegate.Delegate_Selector_placeholder"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
